File: mod/common/java/xen/mod/core/WorldModel.java
package xen.mod.core;

/** Imagination: predicts next observation, reward, harm and the end of a life. Port of world_model.py. */
public final class WorldModel {
	public final Mlp net;
	final int obs, actions;
	public long updates;

	public WorldModel(int obs, int actions, int[] hidden, float lr, long seed) {
		int[] sizes = new int[hidden.length + 2];
		sizes[0] = obs + actions;
		System.arraycopy(hidden, 0, sizes, 1, hidden.length);
		sizes[sizes.length - 1] = obs + 3;
		net = new Mlp(sizes, lr, seed);
		this.obs = obs;
		this.actions = actions;
	}

	float[] input(float[] o, int a) {
		float[] x = new float[obs + actions];
		System.arraycopy(o, 0, x, 0, obs);
		x[obs + a] = 1f;
		return x;
	}

	static float sigmoid(float x) {
		return (float) (1 / (1 + Math.exp(-Math.max(-30, Math.min(30, x)))));
	}

	/** {nextObs..., reward, harm, pDone} for one state and action. */
	public float[] imagine(float[] o, int a) {
		float[] out = net.predict(input(o, a));
		float[] r = new float[obs + 3];
		for (int i = 0; i < obs; i++) r[i] = Math.max(-1f, Math.min(1f, o[i] + out[i]));
		r[obs] = out[obs];
		r[obs + 1] = Math.max(0f, out[obs + 1]);
		r[obs + 2] = sigmoid(out[obs + 2]);
		return r;
	}

	public float surprise(float[] o, int a, float[] next) {
		float[] p = imagine(o, a);
		double s = 0;
		for (int i = 0; i < obs; i++) s += (p[i] - next[i]) * (p[i] - next[i]);
		return (float) (s / obs);
	}

	/** Train on a batch; returns each transition's surprise before learning. */
	public float[] learn(float[][] o, int[] a, float[] reward, float[] harm, float[][] next, float[] done) {
		int n = a.length;
		float[][] x = new float[n][];
		for (int s = 0; s < n; s++) x[s] = input(o[s], a[s]);
		float[][] out = net.forward(x, true);
		float[][] grad = new float[n][obs + 3];
		float[] surprise = new float[n];
		for (int s = 0; s < n; s++) {
			double se = 0;
			for (int i = 0; i < obs; i++) {
				float err = out[s][i] - (next[s][i] - o[s][i]);
				se += err * err;
				grad[s][i] = 2f * err / (n * obs) * 50f;
			}
			surprise[s] = (float) (se / obs);
			grad[s][obs] = Mlp.huber(out[s][obs] - reward[s]) / n;
			grad[s][obs + 1] = Mlp.huber(out[s][obs + 1] - harm[s]) / n;
			grad[s][obs + 2] = (sigmoid(out[s][obs + 2]) - done[s]) / n;
		}
		net.backward(grad);
		updates++;
		return surprise;
	}
}

File: xen/brain/world_model.py
"""Xen's imagination: a learned model of how the world responds to actions.

It predicts the next observation, the reward and the harm of an action.  Its
prediction error is surprise (curiosity), and the thinking cortex uses it to
imagine the consequences of choices before making them.
"""
import numpy as np

from .nn import MLP, huber_grad


def _sigmoid(x):
    return 1.0 / (1.0 + np.exp(-np.clip(x, -30, 30)))


class WorldModel:
    def __init__(self, obs_dim, n_actions, hidden=(256, 256), lr=3e-4, seed=0):
        self.obs_dim = obs_dim
        self.n_actions = n_actions
        self.net = MLP([obs_dim + n_actions, *hidden, obs_dim + 3], lr=lr, seed=seed)
        self.updates = 0

    def _input(self, obs, actions):
        obs = np.atleast_2d(np.asarray(obs, np.float32))
        onehot = np.zeros((len(obs), self.n_actions), np.float32)
        onehot[np.arange(len(obs)), actions] = 1.0
        return np.concatenate([obs, onehot], 1)

    def imagine(self, obs, actions):
        """Predicted (next_obs, reward, harm, p_done) for a batch."""
        obs = np.atleast_2d(np.asarray(obs, np.float32))
        out = self.net.predict(self._input(obs, actions))
        d = self.obs_dim
        next_obs = np.clip(obs + out[:, :d], -1.0, 1.0)
        return next_obs, out[:, d], np.maximum(out[:, d + 1], 0.0), _sigmoid(out[:, d + 2])

    def surprise(self, obs, action, next_obs):
        """How unexpected a single transition was."""
        predicted, _, _, _ = self.imagine(obs, [action])
        return float(np.mean((predicted[0] - next_obs) ** 2))

    def learn(self, obs, actions, reward, harm, next_obs, done):
        """Train on a batch; returns each transition's surprise before learning."""
        out = self.net.forward(self._input(obs, actions))
        d, n = self.obs_dim, len(obs)
        err = out[:, :d] - (next_obs - obs)
        surprise = (err ** 2).mean(1)
        grad = np.empty_like(out)
        grad[:, :d] = 2.0 * err / (n * d) * 50.0
        grad[:, d] = huber_grad(out[:, d] - reward) / n
        grad[:, d + 1] = huber_grad(out[:, d + 1] - harm) / n
        grad[:, d + 2] = (_sigmoid(out[:, d + 2]) - done) / n
        self.net.backward(grad)
        self.updates += 1
        return surprise
